### PR TITLE
feat(ripple): typescript

### PR DESCRIPTION
**Pull request by moog16**
related: #48

### CODE REVIEW

File: packages/ripple/index.tsx [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+    WrappedComponent: React.ComponentType<P>
+  ) => {
+  return class RippledComponent extends React.Component<
+  // Subtract removes any props "InjectedProps" if they are on "P"
+  // This allows the developer to override any props
+  // https://medium.com/@jrwebdev/react-higher-order-component-patterns-in-typescript-42278f7590fb
+    Subtract<P, InjectedProps<Surface, Activator>> & RippledComponentProps<Surface>,
+    RippledComponentState
+    > {
+    foundation_: MDCRippleFoundation | null = null;
```
Comment (patrickrodee) [inline packages/ripple/index.tsx]: Are these private? If so, should we drop the trailing underscore and prefix them with `private`?

Comment (patrickrodee) [reply packages/ripple/index.tsx]: Also, could this be an optional type instead of nullable?

File: packages/ripple/index.tsx [hunk 2]
```diff
@@ -0,0 +1,319 @@
...
+  className: string,
+  onMouseDown: React.MouseEventHandler<T>,
+  onMouseUp: React.MouseEventHandler<T>,
+  onTouchStart: React.TouchEventHandler<T>,
+  onTouchEnd: React.TouchEventHandler<T>,
+  onKeyDown: React.KeyboardEventHandler<T>,
+  onKeyUp: React.KeyboardEventHandler<T>,
+  onFocus: React.FocusEventHandler<T>,
+  onBlur: React.FocusEventHandler<T>,
+  computeBoundingRect?: (surface: T) => {},
```
Comment (patrickrodee) [inline packages/ripple/index.tsx]: Why is this optional?

File: packages/ripple/index.tsx [hunk 3]
```diff
@@ -0,0 +1,319 @@
...
+
+    get style() {
+      const {style: wrappedStyle} = this.props;
+      const {style} = this.state;
+      return Object.assign({}, style, wrappedStyle);
+    }
+
+    render() {
+      const {
+        /* start black list of otherprops */
```
Comment (patrickrodee) [inline packages/ripple/index.tsx]: Nit: use a different name (maybe exclusion list?)

File: packages/ripple/index.tsx (hunk 1 above)
Comment (moog16) [reply packages/ripple/index.tsx]: RE nullable v optional: ya!
RE private: But if I make it private I refer to it in the tests.

Comment (patrickrodee) [reply packages/ripple/index.tsx]: You could declare it as private then add a public getter for both. Would that work or are you setting them in the tests, too?

Comment (moog16) [reply packages/ripple/index.tsx]: in some instances I do set the foundation for stubbing/spying. I know these should be private, so I was thinking we'd do another round of system changes to converting things to be private. Although adding the getter/setter doesn't sound too bad. What are your thoughts?

Comment (patrickrodee) [reply packages/ripple/index.tsx]: I see you also have `initializeFoundation_` and `createAdapter_` which are public but should probably be private, too. Some ideas:
1. Make the `foundation` property configurable via the constructor
2. Make the constructor take an optional factory function that returns an implementation of `MDCRippleFoundation`. You can then update the constructor to set the private foundation property to the return value of the foundation factory OR initialize it with the default behavior.

I can mock up what 2 looks like if you want.

Comment (patrickrodee) [reply packages/ripple/index.tsx]: Ugh I forgot this is React which means modifying the constructor is silly... You could add a `MDCRippleFoundationFactory` field to the `props` that accepts an optional function with the signature `(surface: Surface, activator?: Activator): MDCRippleFoundation`. Then you could pass configure that via props while still having the `foundation` and `unbounded` fields private.

File: packages/ripple/index.tsx [hunk 4]
```diff
@@ -0,0 +1,317 @@
...
+    // surface: This element receives the visual treatment (classes and style) of the ripple.
+    // activator: This element is used to detect whether to activate the ripple. If this is not
+    // provided, the ripple surface will be used to detect activation.
+    initializeFoundation_ = (surface: Surface, activator?: Activator) => {
+      const adapter = this.createAdapter_(surface, activator);
+      this.foundation_ = new MDCRippleFoundation(adapter);
+      this.foundation_.init();
+    };
+
+    createAdapter_ = (surface: Surface, activator?: Activator) => {
```
[the remark this answers is not in the document]
Comment (moog16) [reply packages/ripple/index.tsx]: got a lot of errors when I tried making them private :(

**Comment by moog16** [comment]
@amrtgaber & @patrickrodee Thanks for the initial reviews. I updated according to your requested changes. Please let me know what you think. Will be online in the afternoon.

**Pull request merged**